SEP-0010: Add web auth domain manage data operation (v3.1.0) (#767)
stellar/stellar-protocol: ecosystem/sep-0010.md

This protocol is a variation of mutual challenge-response, which uses Stellar transactions to encode challenges and responses.

It involves three components:
- A domain hosting a [SEP-1 stellar.toml](sep-0001.md) containing a `WEB_AUTH_ENDPOINT` (URL) and `SIGNING_KEY` (`G...`), referred to as the home domain.

> - An endpoint providing the GET and POST operations discussed in this document. The endpoint may be hosted on the home domain, a sub-domain of the home domain, or any other domain. The domain that it is hosted is referred to as the web auth domain. 

Reviewer: This last sentence doesn't read well, `The domain that it is hosted`?

Author: ```suggestion
- An endpoint providing the GET and POST operations discussed in this document. The endpoint may be hosted on the home domain, a sub-domain of the home domain, or any other domain. The domain that the endpoint is hosted is referred to as the web auth domain. 
```

Reviewer: Added `on`, LGTM now

   1. Source account set to the user's account.
   1. Key set to `<home domain> auth` where the home domain is the home domain from the discovery flow.
   1. Value set to a nonce value.

> 1. The client verifies that if the transaction has a Manage Data operation with key `server_domain` that it has:

Reviewer: I'm a bit concerned that `server_domain` is too vague. What about `sep10_server_domain`?

Author: I think we should keep the term `SEP-10` out of the code interface, SEPs do not normally include that. How about `issuer_domain`? Using this name will do one of two things:
- Clear up misconception because the issuer of the SEP-10 challenge is the SEP-10 server.
- Any ambiguity should be more apparent since the term `issuer` may more likely lead to the reader digging in to find out what it means rather than glossing over on a generic term such as `server`.

Reviewer: I think `server_domain` is a better than `issuer_domain` due to the fact that the term `issuer` is already a keyword within the Stellar context and could confuse the reader. What about `web_auth_server`, which would be the domain used in SEP-1's `WEB_AUTH_ENDPOINT`?

Author: Oh I like `web_auth_domain`. It correlates directly to this SEP. I like it. Updated. 01a77b4

Author: One more thought on this though, we use the term `server_account`  to mean the account of the SEP-10 server, so if we continued to use the term `server_domain` it correlates.

Reviewer: Ah thats true, that definitely makes it more plausible. I still think `web_auth_domain` is better though.

    * `manage_data(source: client_account, key: '<home domain> auth', value: random_nonce())`

>       * The value of key is the home domain the challenge is authenticating, followed by `auth`. It can be at most 64 characters.

Reviewer: `the home domain the challenge is authenticating`

The challenge does not authenticate a home domain. I think we should change this to either mention the discovery flow like in prior statements, or we should be explicit about what domain we're talking about.

```suggestion
      * The value of key is the domain hosting the service's [SEP-1 stellar.toml](sep-0001.md), followed by `auth`. It can be at most 64 characters.
```

Author: The home domain is defined at the top of the document, so that we do not need to redefine it at each mention point throughout the document.

>The challenge does not authenticate a home domain.

I agree, and I think there's a word missing here. It should say `the home domain the challenge is authenticating [with]...`.

I think this is out-of-scope of this PR. If you agree, can you open a new PR with the improvements you propose?

      * The value must be 64 bytes long. It contains a 48 byte cryptographic-quality random string encoded using base64 (for a total of 64 bytes after encoding).

>     * subsequent operations, order unimportant:

Reviewer: ```suggestion
    * subsequent operations must conform to one of the following patterns listed below. The order of these `manage_data` does not matter.
```

Author: The statement `must conform to one of the following patterns` reads to me as if the server must choose one dot point and only implement that, which isn't the intention. The server does not need to choose only one of the listed patterns, it can include any of the data that is listed, and the order with which it includes them in the transaction is not important.

It's unclear from your comment what is unclear about the current text that you're wanting to make clear. Could you elaborate so I understand the problem you're solving with the suggestion?

Reviewer: It was really a matter of verbosity, we can keep it as-is but I felt the current language was a bit too terse.

Author: Agreed, it's a bit terse. I'm struggling to come up with more words that don't raise other questions so I'm going to leave it for now.
